server/paymentProviders/stripe/webhook.ts

```diff
@@ -92,6 +92,55 @@ async function createOrUpdateOrderStripePaymentMethod(
   return pm;
 }
 
+export const mandateUpdated = async (event: Stripe.Event) => {
+  const stripeAccount = event.account ?? config.stripe.accountId;
+
+  const stripeMandate = event.data.object as Stripe.Mandate;
+
+  const stripePaymentMethodId =
+    typeof stripeMandate.payment_method === 'string' ? stripeMandate.payment_method : stripeMandate.payment_method.id;
+
+  const paymentMethod = await models.PaymentMethod.findOne({
+    where: {
+      data: {
+        stripePaymentMethodId,
+      },
+    },
+  });
+
+  if (!paymentMethod) {
+    const stripePaymentMethod = await stripe.paymentMethods.retrieve(stripePaymentMethodId, {
+      stripeAccount,
+    });
+
+    await models.PaymentMethod.create({
+      name: formatPaymentMethodName(stripePaymentMethod),
+      service: PAYMENT_METHOD_SERVICE.STRIPE,
+      type: stripePaymentMethod.type,
+      confirmedAt: new Date(),
+      saved: stripeMandate.type === 'multi_use' && stripeMandate.status !== 'inactive',
+      data: {
+        stripePaymentMethodId: stripePaymentMethod.id,
+        stripeAccount,
+        ...mapStripePaymentMethodExtraData(stripePaymentMethod),
+        stripeMandate,
+      },
+    });
```

Comment:
@kewitz I found while testing this that the `mandate.updated` event is received before a payment method is attached to a customer. So at this time it might be that the payment method does not exist on our side and so we lose the information about the mandate.

I have changed here to create a payment method when we first see it on the `mandate.updated` event, and when we receive the `payment_method.attached` we save the payment method to the customer.

Reply:
We seem to have a race condition between `mandate.update` and `payment_method.attached` since Stripe doesn't guarantee the order of events sent, nor can we guarantee the speed at which we return these functions.

I suggest wrapping both handlers in transactions. We can also use the method `findOrCreate` to simplify the first conditional of these handlers.